colexec: populate batches for projecting operators upfront

yuzefovich opened the pull request:
**colexec: populate batches for projecting operators upfront**

Release justification: bug fix (without this, vectorized engine could
hit internal errors in some cases).

This commit introduces two utility operators `vectorTypeEnforcer` and
`batchSchemaPrefixEnforcer` which all "projecting" operators that do not
output their own batches need to use. This utility operators are planned
in a such way that they wrap the input to the projecting operator and
make sure that there is an output column of the appropriate type for
corresponding projecting operator. For most operators, enforcing
a single column is sufficient, however, `caseOp`, `andProjOp`, and
`orProjOp` use side chains of projecting operators, so they need to
enforce the vector types on the prefix of the columns from the batch, so
they use `batchSchemaPrefixEnforcer`. Such design allows us to have
short-circuiting logic in all operators on zero-length batches, yet it
enforces that all vectors are correctly appended upfront.

Additionally, this commit refactors several unit tests to use
`NewColOperator` to plan the operators to test rather than instantiating
them directly. This is done so that planning code is being tested as
well, and it already found a deficiency in planning casts to decimals
(namely, currently we're prohibiting all possible casts except between
decimals of the same precision).

Fixes: #46646.

Release note: None

**colexec: fix width mismatch for integer types in some cases**

We have a limitation of our SQL type system that it doesn't respect the
width of integer columns, however, the vectorized projecting operators
do, and this can lead to a mismatch between expected logical and actual
physical types in the vectorized flow. This is now fixed by planning
a cast that "synchronizes" the reality and the expectations.

Fixes: #46714.

Release note (bug fix): Previously, CockroachDB could hit an internal
error when queries with projections only of `INT2` and/or `INT4` columns
were executed via the vectorized engine, and this has been fixed.

yuzefovich committed:
colexec: populate batches for projecting operators upfront

Release justification: bug fix (without this, vectorized engine could
hit internal errors in some cases).

This commit introduces two utility operators `vectorTypeEnforcer` and
`batchSchemaPrefixEnforcer` which all "projecting" operators that do not
output their own batches need to use. This utility operators are planned
in a such way that they wrap the input to the projecting operator and
make sure that there is an output column of the appropriate type for
corresponding projecting operator. For most operators, enforcing
a single column is sufficient, however, `caseOp`, `andProjOp`, and
`orProjOp` use side chains of projecting operators, so they need to
enforce the vector types on the prefix of the columns from the batch, so
they use `batchSchemaPrefixEnforcer`. Such design allows us to have
short-circuiting logic in all operators on zero-length batches, yet it
enforces that all vectors are correctly appended upfront. This also
allows us to change the behavior of `maybeAddColumn` to never append
"unknown" vectors because planning code is now responsible for setting
up type enforcers so that only well-typed columns are appended.

Additionally, this commit refactors several unit tests to use
`NewColOperator` to plan the operators to test rather than instantiating
them directly. This is done so that planning code is being tested as
well.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46712)
<!-- Reviewable:end -->

yuzefovich commented:
To see what is the impact on the performance, I ran all TPC-H queries that use `caseOp`, and It seems like there is probably a minor hit, borderline with noise. The queries were run in three configs (master, master + this PR, master + this PR + `caseOp` disabled).

*before*
```
I200328 23:49:04.054963 83 workload/tpch/tpch.go:465  [q7] returned 4 rows after 6.11 seconds
I200328 23:49:09.815813 83 workload/tpch/tpch.go:465  [q7] returned 4 rows after 5.76 seconds
I200328 23:49:15.790034 83 workload/tpch/tpch.go:465  [q7] returned 4 rows after 5.97 seconds
```
*after*
```
I200328 23:46:43.948210 25 workload/tpch/tpch.go:465  [q7] returned 4 rows after 5.92 seconds
I200328 23:46:49.705242 25 workload/tpch/tpch.go:465  [q7] returned 4 rows after 5.76 seconds
I200328 23:46:55.656657 25 workload/tpch/tpch.go:465  [q7] returned 4 rows after 5.95 seconds
```
*`caseOp` disabled*
```
I200328 23:59:53.006189 40 workload/tpch/tpch.go:465  [q7] returned 4 rows after 6.02 seconds
I200328 23:59:58.817386 40 workload/tpch/tpch.go:465  [q7] returned 4 rows after 5.81 seconds
I200329 00:00:04.838425 40 workload/tpch/tpch.go:465  [q7] returned 4 rows after 6.02 seconds
```
---
*before*
```
I200328 23:49:30.973944 83 workload/tpch/tpch.go:465  [q8] returned 2 rows after 11.10 seconds
I200328 23:49:41.836040 83 workload/tpch/tpch.go:465  [q8] returned 2 rows after 10.86 seconds
I200328 23:49:52.741194 83 workload/tpch/tpch.go:465  [q8] returned 2 rows after 10.91 seconds
```
*after*
```
I200328 23:47:14.091102 40 workload/tpch/tpch.go:465  [q8] returned 2 rows after 12.10 seconds
I200328 23:47:25.305368 40 workload/tpch/tpch.go:465  [q8] returned 2 rows after 11.21 seconds
I200328 23:47:36.380799 40 workload/tpch/tpch.go:465  [q8] returned 2 rows after 11.08 seconds
```
*`caseOp` disabled*
```
I200329 00:00:18.442021 38 workload/tpch/tpch.go:465  [q8] returned 2 rows after 11.36 seconds
I200329 00:00:29.756663 38 workload/tpch/tpch.go:465  [q8] returned 2 rows after 11.31 seconds
I200329 00:00:40.888328 38 workload/tpch/tpch.go:465  [q8] returned 2 rows after 11.13 seconds
```
---
*before*
```
I200328 23:49:56.181440 39 workload/tpch/tpch.go:465  [q14] returned 1 rows after 0.96 seconds
I200328 23:49:57.054479 39 workload/tpch/tpch.go:465  [q14] returned 1 rows after 0.87 seconds
I200328 23:49:57.919173 39 workload/tpch/tpch.go:465  [q14] returned 1 rows after 0.86 seconds
```
*after*
```
I200328 23:47:41.615062 29 workload/tpch/tpch.go:465  [q14] returned 1 rows after 1.00 seconds
I200328 23:47:42.577215 29 workload/tpch/tpch.go:465  [q14] returned 1 rows after 0.96 seconds
I200328 23:47:43.625086 29 workload/tpch/tpch.go:465  [q14] returned 1 rows after 1.05 seconds
```
*`caseOp` disabled*
```
I200329 00:00:44.445179 44 workload/tpch/tpch.go:465  [q14] returned 1 rows after 1.06 seconds
I200329 00:00:45.433332 44 workload/tpch/tpch.go:465  [q14] returned 1 rows after 0.99 seconds
I200329 00:00:46.427353 44 workload/tpch/tpch.go:465  [q14] returned 1 rows after 0.99 seconds
```
---
*before*
```
I200328 23:50:06.859247 42 workload/tpch/tpch.go:465  [q19] returned 1 rows after 4.82 seconds
I200328 23:50:11.481752 42 workload/tpch/tpch.go:465  [q19] returned 1 rows after 4.62 seconds
I200328 23:50:16.273733 42 workload/tpch/tpch.go:465  [q19] returned 1 rows after 4.79 seconds
```
*after*
```
I200328 23:47:52.752876 24 workload/tpch/tpch.go:465  [q19] returned 1 rows after 5.05 seconds
I200328 23:47:57.567305 24 workload/tpch/tpch.go:465  [q19] returned 1 rows after 4.81 seconds
I200328 23:48:02.451034 24 workload/tpch/tpch.go:465  [q19] returned 1 rows after 4.88 seconds
```
*`caseOp` disabled*
```
I200329 00:00:54.387985 66 workload/tpch/tpch.go:465  [q19] returned 1 rows after 5.06 seconds
I200329 00:00:59.472163 66 workload/tpch/tpch.go:465  [q19] returned 1 rows after 5.08 seconds
I200329 00:01:04.598006 66 workload/tpch/tpch.go:465  [q19] returned 1 rows after 5.13 seconds
```

So the current solution looks acceptable.

asubiotto reviewed (changes requested):
Nice to see this done so quickly.

> Reviewed 30 of 30 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-:-M3f1Pp8D1CD5jK5A7hX:bvijf9k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/allocator.go, line 166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3eqbK42gAimgs-it-5:-M3eqbK42gAimgs-it-6:b966kls) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/allocator.go#L166)):*
> ```Go
> // b at position colIdx. It will do so if either
> // 1. the width of the batch is not greater than colIdx, or
> // 2. there is already an "unknown" vector in position colIdx in the batch.
> ```

This condition exists because of previous behavior where we would want to handle operators not appending columns. Do we still need it given this new behavior of always enforcing a schema? It might be prudent to keep it in.
___
*[pkg/sql/colexec/execplan.go, line 1545 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3erNuS-lGSiwNk1O49:-M3erNuS-lGSiwNk1O4A:b-j0h7a4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/execplan.go#L1545)):*
> ```Go
> 	switch toType.Family() {
> 	case types.DecimalFamily:
> 		// TODO(yuzefovich): do we want to prohibit casts to decimals *only*
> ```

The problem is that if the decimal `toType` is non-default, the cast could be wrong. We could allow casts to the default decimal type (with `Width` and `Precision` 0), but not sure we want to do that right now.
___
*[pkg/sql/colexec/operator.go, line 293 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3euFj67euSaV85-V_G:-M3euFj72jiR-tMDnQ_C:b-v3rzwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/operator.go#L293)):*
> ```Go
> // enforces that non-zero length batch from the input has a vector of the
> // desired type in the desired position. If the width of the batch is less than
> // the desired position, than new vector will be appended; if the batch has
> ```

`s/than new/then a new || a new/`
___
*[pkg/sql/colexec/operator.go, line 294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3euLRG0Ka69AT3L3sW:-M3euLRG0Ka69AT3L3sX:b-dpg81r) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/operator.go#L294)):*
> ```Go
> // desired type in the desired position. If the width of the batch is less than
> // the desired position, than new vector will be appended; if the batch has
> // well-typed vector of undesired type in the desired position, an error will
> ```

`s/well-typed/a well-typed`, `s/of undesired/of an undesired`
___
*[pkg/sql/colexec/operator.go, line 300 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3euVnA-4xUNPFWRmyi:-M3euVnA-4xUNPFWRmyj:b-5cbz4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/operator.go#L300)):*
> ```Go
> // "projecting" Operator (such Operator that has a single column as its output
> // and does not touch other columns by simply passing them along). The type
> // schema passed in into vectorTypeEnforcer *must* include the output
> ```

`s/in into/into`
___
*[pkg/sql/colexec/operator.go, line 327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3f0mv5EkNemvGFFDaS:-M3f0mv5EkNemvGFFDaT:bwefs1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/operator.go#L327)):*
> ```Go
> var _ Operator = &vectorTypeEnforcer{}
> 
> func newVectorTypeEnforcer(allocator *Allocator, input Operator, typ coltypes.T, idx int) Operator {
> ```

Should these be in a different file?
___
*[pkg/sql/colexec/operator.go, line 351 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3eupQr9Lp_41_O5JP_:-M3eupQr9Lp_41_O5JPa:b-xj6pp0) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/operator.go#L351)):*
> ```Go
> // batchSchemaPrefixEnforcer is similar to vectorTypeEnforcer in its purpose,
> // but it enforces that the prefix of the columns of the non-zero length batch
> // satisfies the desired schema. It needs to wrap the input to a "projecting"
> ```

to a "projecting" (operator?)


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46712)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/allocator.go, line 166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3eqbK42gAimgs-it-5:-M3hOPDgAkyg9QEcR6vV:bdd1aqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/allocator.go#L166)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This condition exists because of previous behavior where we would want to handle operators not appending columns. Do we still need it given this new behavior of always enforcing a schema? It might be prudent to keep it in.
</blockquote></details>

You're right, this is no longer needed, in fact, having a wrong-type vector should be an error. Updated.

However, in regards to removing a map from `simpleProjectOp` - we cannot do that. The problem described in #45686 is still present after this refactor (`projectingBatch`'s `projection` field will incorrectly tell us the width of the batch when a different "internally" batch appears).
___
*[pkg/sql/colexec/execplan.go, line 1545 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3erNuS-lGSiwNk1O49:-M3gx5n6-eVSaX6aRD3o:b-uhyz7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/execplan.go#L1545)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

The problem is that if the decimal `toType` is non-default, the cast could be wrong. We could allow casts to the default decimal type (with `Width` and `Precision` 0), but not sure we want to do that right now.
</blockquote></details>

Ok, left a clarifying comment. I thought that we didn't allow the casts only from a decimal to another decimal with different precision, so this was unexpected to me.
___
*[pkg/sql/colexec/operator.go, line 293 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3euFj67euSaV85-V_G:-M3gxSB98MKrmxZCPOaO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/operator.go#L293)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

`s/than new/then a new || a new/`
</blockquote></details>

Done.
___
*[pkg/sql/colexec/operator.go, line 294 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3euLRG0Ka69AT3L3sW:-M3gxSge-i2OWO9Gy85R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/operator.go#L294)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

`s/well-typed/a well-typed`, `s/of undesired/of an undesired`
</blockquote></details>

Done.
___
*[pkg/sql/colexec/operator.go, line 300 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3euVnA-4xUNPFWRmyi:-M3gxZU17twuERQc6Thm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/operator.go#L300)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

`s/in into/into`
</blockquote></details>

Done.
___
*[pkg/sql/colexec/operator.go, line 327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3f0mv5EkNemvGFFDaS:-M3gxggw-3iywuL_R3BD:bhhl833) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/operator.go#L327)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Should these be in a different file?
</blockquote></details>

I don't see a better file for it (other than creating a new `projecting_ops_util.go` or something like that) since these operators are reused by many different projecting operators.
___
*[pkg/sql/colexec/operator.go, line 351 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3eupQr9Lp_41_O5JP_:-M3gy0SwBJAYBiAEx2ZS:b-9f087b) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/operator.go#L351)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

to a "projecting" (operator?)
</blockquote></details>

Yep, done.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
colexec: fix width mismatch for integer types in some cases

We have a limitation of our SQL type system that it doesn't respect the
width of integer columns, however, the vectorized projecting operators
do, and this can lead to a mismatch between expected logical and actual
physical types in the vectorized flow. This is now fixed by planning
a cast that "synchronizes" the reality and the expectations.

Release note (bug fix): Previously, CockroachDB could hit an internal
error when queries with projections only of `INT2` and/or `INT4` columns
were executed via the vectorized engine, and this has been fixed.

asubiotto reviewed (changes requested):
> Reviewed 11 of 11 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46712)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/execplan.go, line 1828 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3jyTpP2cVNttLSUMy6:-M3jyTpP2cVNttLSUMy7:b-5kaahs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5721afb0f9306bcd836e1a8764fe84664b10d41/pkg/sql/colexec/execplan.go#L1828)):*
> ```Go
> 		// and if so, we override the output physical type to be what the
> 		// vectorized projection operators expect.
> 		leftPhysType := typeconv.FromColumnType(left.ResolvedType())
> ```

Does this belong in this commit?
___
*[pkg/sql/colexec/execplan.go, line 1830 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3jyX3b5whMkFSDzU3p:-M3jyX3b5whMkFSDzU3q:byhqocy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5721afb0f9306bcd836e1a8764fe84664b10d41/pkg/sql/colexec/execplan.go#L1830)):*
> ```Go
> 		leftPhysType := typeconv.FromColumnType(left.ResolvedType())
> 		rightPhysType := typeconv.FromColumnType(right.ResolvedType())
> 		if leftPhysType == coltypes.Int16 && rightPhysType == coltypes.Int16 {
> ```

What about int8? I have to admit I'm not 100% confident in this type promotion or my understanding of it. Is there some pre-existing code that determines the output column type from `left.ResolvedType()` and `right.ResolvedType()` that we can then convert to a physical type?
___
*[pkg/sql/colexec/operator.go, line 327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3f0mv5EkNemvGFFDaS:-M3k3QFwD9gHFcdl3tXn:b-8glma8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/operator.go#L327)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I don't see a better file for it (other than creating a new `projecting_ops_util.go` or something like that) since these operators are reused by many different projecting operators.
</blockquote></details>

Yeah, I guess I was referring to making a different file but it's not a big deal.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46712)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @otan)

___
*[pkg/sql/colexec/execplan.go, line 1828 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3jyTpP2cVNttLSUMy6:-M3kvS6T44BMu1q47aoI:b25czp5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5721afb0f9306bcd836e1a8764fe84664b10d41/pkg/sql/colexec/execplan.go#L1828)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Does this belong in this commit?
</blockquote></details>

Without the second commit, `BenchmarkProjOp` (which has been refactored to use the planning code) encounters an internal error when using `Int32` as the input type and crashes. I guess you could argue that the change to the benchmark itself doesn't belong in the first commit either, but I think that eventually we want to be in a position when all tests live `colexec_test` package and most unit tests would use `colexec.NewColOperator` to create operators to test. The only exception would be some tests that enforce that a particular operator is instantiated.

Anyway, I encountered this problem with mismatched width while working on a single issue, so I'm inclined to keep everything in a single commit.
___
*[pkg/sql/colexec/execplan.go, line 1830 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3jyX3b5whMkFSDzU3p:-M3kwxLZ8tOTvLCtVqDh:bpx46xk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5721afb0f9306bcd836e1a8764fe84664b10d41/pkg/sql/colexec/execplan.go#L1830)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

What about int8? I have to admit I'm not 100% confident in this type promotion or my understanding of it. Is there some pre-existing code that determines the output column type from `left.ResolvedType()` and `right.ResolvedType()` that we can then convert to a physical type?
</blockquote></details>

"What about int8" - do you mean why we not have any cases with `coltypes.Int64`? Because in those cases we already have the correct output logical and physical type.

My understanding is that our logical type system assumes everywhere that we operate with `INT8` when operating with any integers (so there are implicit casts). The place where the return type for binary expressions is defined is `tree/eval.go`, and there is only a single overload for `+` operating on integers:
```
	Plus: {
		&BinOp{
			LeftType:   types.Int,
			RightType:  types.Int,
			ReturnType: types.Int,
			Fn: func(_ *EvalContext, left Datum, right Datum) (Datum, error) {
				a, b := MustBeDInt(left), MustBeDInt(right)
				r, ok := arith.AddWithOverflow(int64(a), int64(b))
				if !ok {
					return nil, ErrIntOutOfRange
				}
				return NewDInt(DInt(r)), nil
			},
		},
```
so AFAIU it is safe to assume (at the moment) that we only need to check whether SQL type is `INT8`, and if so, do some magic to make an "adaptation" between SQL type system's assumptions and vectorized engine's expectations.

By the way, I think we can have non-`INT8` integer columns on the left or right side only when the input is populated by the vectorized engine. In theory, we could change `cFetcher` to do the upcast to `INT8` of `INT2` and `INT4`, and then we would not need to have this custom logic.

However, the root of the problem lies outside of the vectorized engine, but I believe that we have plans to fix the resolution of overloads in the type checker. cc @otan

cc @jordanlewis on the general discussion here 
___
*[pkg/sql/colexec/operator.go, line 327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3f0mv5EkNemvGFFDaS:-M3kzTZI2a-UPaAiVSfN:b-cfx66j) ([raw file](https://github.com/cockroachdb/cockroach/blob/2426516dd8313e9215515950fc1f35145c535896/pkg/sql/colexec/operator.go#L327)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Yeah, I guess I was referring to making a different file but it's not a big deal.
</blockquote></details>

Ok, I'll keep it in here then.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46712)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @otan)

___
*[pkg/sql/colexec/execplan.go, line 1830 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3jyX3b5whMkFSDzU3p:-M3lXmVC1wPS5-yLdJMJ:bxy3b7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5721afb0f9306bcd836e1a8764fe84664b10d41/pkg/sql/colexec/execplan.go#L1830)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

"What about int8" - do you mean why we not have any cases with `coltypes.Int64`? Because in those cases we already have the correct output logical and physical type.

My understanding is that our logical type system assumes everywhere that we operate with `INT8` when operating with any integers (so there are implicit casts). The place where the return type for binary expressions is defined is `tree/eval.go`, and there is only a single overload for `+` operating on integers:
```
	Plus: {
		&BinOp{
			LeftType:   types.Int,
			RightType:  types.Int,
			ReturnType: types.Int,
			Fn: func(_ *EvalContext, left Datum, right Datum) (Datum, error) {
				a, b := MustBeDInt(left), MustBeDInt(right)
				r, ok := arith.AddWithOverflow(int64(a), int64(b))
				if !ok {
					return nil, ErrIntOutOfRange
				}
				return NewDInt(DInt(r)), nil
			},
		},
```
so AFAIU it is safe to assume (at the moment) that we only need to check whether SQL type is `INT8`, and if so, do some magic to make an "adaptation" between SQL type system's assumptions and vectorized engine's expectations.

By the way, I think we can have non-`INT8` integer columns on the left or right side only when the input is populated by the vectorized engine. In theory, we could change `cFetcher` to do the upcast to `INT8` of `INT2` and `INT4`, and then we would not need to have this custom logic.

However, the root of the problem lies outside of the vectorized engine, but I believe that we have plans to fix the resolution of overloads in the type checker. cc @otan

cc @jordanlewis on the general discussion here 
</blockquote></details>

I guess I'm more wondering why the cases don't have a reference to `coltypes.Int8`.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
> I guess I'm more wondering why the cases don't have a reference to `coltypes.Int8`.

I think you meant `coltypes.Int64`, and I tried to answer it here:
> By the way, I think we can have non-INT8 integer columns on the left or right side only when the input is populated by the vectorized engine.

What happens is that `cFetcher` sees columns of, say, `INT4`, so it populates its batch with `coltypes.Int32` vector (i.e. `cFetcher` "respects" the width of integer types), and then projecting operator sees it as well as `coltypes.Int32` and will output also a vector of `coltypes.Int32` (i.e. it also "respects" the width). However, SQL type system does *not* respect the width, so we have to have this custom overriding logic in place.

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-:-M3lbluE88QisJ2p_fij:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @otan)

___
*[pkg/sql/colexec/execplan.go, line 1830 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3jyX3b5whMkFSDzU3p:-M3lbWgAEihpTZO3RU9k:bxsshyd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5721afb0f9306bcd836e1a8764fe84664b10d41/pkg/sql/colexec/execplan.go#L1830)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I guess I'm more wondering why the cases don't have a reference to `coltypes.Int8`.
</blockquote></details>

Gotcha, all these `Int<N>s are confusing me.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46712)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @jordanlewis, and @otan)

___
*[pkg/sql/colexec/execplan.go, line 1830 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3jyX3b5whMkFSDzU3p:-M3re-fS8U4YXZhMbJvz:bd1zvlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5721afb0f9306bcd836e1a8764fe84664b10d41/pkg/sql/colexec/execplan.go#L1830)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Gotcha, all these `Int<N>s are confusing me.
</blockquote></details>

Actually, my initial solution was incorrect (it would fail if we had a type mismatch at the "middle" materializer when we're wrapping rowexec processors), and I updated the PR to plan casts. This solution seems solid. PTAL.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r4, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-:-M3u9akj9QDmfFTQz-Fw:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @otan)



<!-- Sent from Reviewable.io -->

otan commented:
Can confirm its safe to use int8 for all type checking goodness for now.
But ideally you use the exact type if you can - resolution should pick it
up anyway.

On Thu, 2 Apr 2020, 2:13 am Alfonso Subiotto Marqués, <
notifications@github.com> wrote:

> *@asubiotto* approved this pull request.
>
> [image: :lgtm:]
> <https://camo.githubusercontent.com/41b3ec3419116e3b3f84507ad965646ce4ce1f0c/68747470733a2f2f72657669657761626c652e696f2f6c67746d2e706e67>
>
> Reviewed 3 of 3 files at r4, 2 of 2 files at r5.
> *Reviewable
> <https://reviewable.io/reviews/cockroachdb/cockroach/46712#-:-M3u9akj9QDmfFTQz-Fw:bg7igkv>*
> status: [image: :shipit:] complete! 1 of 0 LGTMs obtained (waiting on
> @jordanlewis <https://github.com/jordanlewis> and @otan
> <https://github.com/otan>)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/46712#pullrequestreview-386242769>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA32FQ4SZGIEXM27CKLXCKTRKRJL3ANCNFSM4LVWDKRQ>
> .
>

jordanlewis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46712)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @otan, and @yuzefovich)

___
*[pkg/sql/colexec/execplan.go, line 1946 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3wDlIN-rknTVvzimEu:-M3wDlIN-rknTVvzimEv:b-6l61av) ([raw file](https://github.com/cockroachdb/cockroach/blob/d801cc9894b913fccf2c5386784c475d4ada2f0d/pkg/sql/colexec/execplan.go#L1946)):*
> ```Go
> 		// The projection operator outputs a column of a different type than
> 		// the expected logical type. In order to "synchronize" the reality and
> 		// the expectations, we plan a cast.
> ```

Please add a concrete example here. Specifically, I think we should keep careful track of why this is here and remove it later. We only think it's a problem for integer widths right now, right?
___
*[pkg/sql/colexec/operator.go, line 354 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3wK9L14-71qeSNaJ6q:-M3wK9L14-71qeSNaJ6r:b6vvb54) ([raw file](https://github.com/cockroachdb/cockroach/blob/d801cc9894b913fccf2c5386784c475d4ada2f0d/pkg/sql/colexec/operator.go#L354)):*
> ```Go
> //
> // NOTE: the type schema passed into batchSchemaPrefixEnforcer *must* include
> // the output type of the Operator that the enforcer will be the input to.
> ```

I don't really love this, because it's not obvious that the right operators will remember to use this thing at instantation. But, it's okay for now. I think we should do better here over time though if we can.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-:-M3wVJj253ThozaiiQ-O:b-tlx7ko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @jordanlewis, and @otan)

___
*[pkg/sql/colexec/execplan.go, line 1946 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3wDlIN-rknTVvzimEu:-M3wTvpA9EnVyZR78Vge:b-mz8qhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d801cc9894b913fccf2c5386784c475d4ada2f0d/pkg/sql/colexec/execplan.go#L1946)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Please add a concrete example here. Specifically, I think we should keep careful track of why this is here and remove it later. We only think it's a problem for integer widths right now, right?
</blockquote></details>

Added a comment and filed an issue.
___
*[pkg/sql/colexec/operator.go, line 354 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46712#-M3wK9L14-71qeSNaJ6q:-M3wU7U037A1MCisCUR-:bpra9rq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d801cc9894b913fccf2c5386784c475d4ada2f0d/pkg/sql/colexec/operator.go#L354)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I don't really love this, because it's not obvious that the right operators will remember to use this thing at instantation. But, it's okay for now. I think we should do better here over time though if we can.
</blockquote></details>

Hm, I agree this is not perfect, but I don't think this operator will necessarily be a source of bugs - we now explicitly prohibit `AppendCol` in almost all files, so when writing a new `Operator` that needs to append a column (or multiple columns) we will be forced to use either `vectorTypeEnforcer` or `batchSchemaPrefixEnforcer`. If we mistakenly use the former when we need the latter or use the latter but with incorrect prefix, then the unit tests *should* fail (as long as we're reusing the same planning code).


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Noticed a typo in the comment.

bors r-

yuzefovich commented:
bors r+

yuzefovich commented:
bors r+

yuzefovich commented:
I had to make a change to support `coltypes.Unhandled` in the type schema prefix. PTAL.

asubiotto reviewed (approved):
> Reviewed 3 of 3 files at r6, 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46712)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
bors r+
